src/components/Modal/Modal.unit.spec.jsx

```diff
@@ -100,11 +102,82 @@ describe('<Modal/>', () => {
           <button id="first-button">First button</button>
           <button id="second-button">Second button</button>
         </Modal>,
-        { attachTo: root }
+        { attachTo: root },
       );
 
       expect(global.document.activeElement.id).to.equal('second-button');
       tree.unmount();
     });
   });
+
+  describe('analytics event', function () {
+    it('should be triggered when modal is visible', () => {
+      const handleAnalyticsEvent = e => {
+        expect(e.detail).to.eql({
+          componentName: 'Modal',
+          action: 'show',
+          details: {
+            status: undefined,
+            title: undefined,
+            primaryButtonText: undefined,
+            secondaryButtonText: undefined,
+          },
+        });
+      };
+
+      global.document.body.addEventListener(
+        'component-library-analytics',
+        handleAnalyticsEvent,
+      );
+
+      mount(
+        <Modal id="modal" visible={true} clickToClose onClose={() => {}}>
+          Modal contents
+        </Modal>,
+      );
+
+      global.document.body.removeEventListener(
+        'component-library-analytics',
+        handleAnalyticsEvent,
+      );
+    });
```

Comment:
Is this the best way to test these events?  Basically recreate the listener and look at `e.details`?

Or should we be using sinon.spy or something?

Reply:
I think we can get away with not using `sinon.spy` here. Still, as it's written right now, this test will pass if `handleAnalyticsEvents` is never called.

The general approach of creating an event listener seems okay to me, but I have concerns that timing might get...weird. Like, if the test ends (because execution ends) before the event listener is called. My suspicion is the [execution phase](https://medium.com/@nazanindelam/all-you-need-to-learn-to-understand-javascript-execution-context-and-stack-3babbdd88868) may sometimes happen like so:
1. The event listener call will get queued up in the JS event loop
2. The `it()` test invocation will finish
3. The event loop will pop off the next event (calling the handler)...after our test has completed

(Plug for [this great talk](https://www.youtube.com/watch?v=8aGhZQkoFbQ) on how it all works together in JS.)

I don't know if that'll _actually_ happen, but I think we need to be careful here. I'll pull this branch down and run some tests on it myself to see if I can come up with a clever way to ensure that our event listener is called (if it's going to be called) before the end of the test while also ensuring that we don't wait around for some arbitrary amount of time.

Right now, I'm leaning toward grabbing the `done` callback in the test function definition, pulling up the `handleAnalyticsEvents` into its own function which returns a closure that runs that `expect` against an object we pass it, and then calls the `done` callback...which sounds way more complicated with English than it does with code. :dizzy_face:

Reply:
Well good news! I can't get this test to fail your way. :smile: 

I tried it the way I was thinking and it also works, but I realized by calling `done()` in the event listener callback, I can't then remove the event listener from the DOM, so that's no good. I mean, it may be fine if each test runs with a new DOM, but I'm not super confident that that happens, so :man_shrugging:.

If we find this produces flaky tests somewhere down the line, we can reassess.

Reply:
OK cool - I'll be writing a ton of these tests so I will keep an eye on things.